### ISSUE TITLE
Parameter pack causes crash when generic types are initialized concurrently

### ISSUE THREAD

**Issue by ThyOwen** [apple/swift#72480]
### Description

In the function below, a task group is initialized to search multiple webpages at the same time (1 webpage -> repeat each Multiple return types)
there is also a simplified version which doesn't return the bad memory address.

the inferred types of the two differ:
Parameter pack version: `[((GenericWebpage)?, (GenericWebpage)?)]`
Simplified version: `[(GenericWebpage?, GenericWebpage?)]`

if I change from class to struct, mark `mutating get()` function all is fixed. Marking `@MainActor` also doesn't address the trouble 

### Reproduction

```swift

public protocol SearchResult {
    init(doc : Document?) throws
}

public struct GenericWebpage : SearchResult {
    let text : String?

    public init(doc : Document?) throws {
        self.text = try? doc?.text() ?? ""
    }
}

@MainActor
public class Search<SearchURLType : SearchURL> {

    typealias SearchParamsType = SearchURLType.SearchParamsType //associated types from other protocals
    typealias SearchItemType = SearchURLType.SearchItemType //associated types from other protocals

    let searchParams : SearchParamsType
    let searchURL : SearchURLType
    var searchItems : [SearchItemType]
    
    init(_ params : SearchParamsType) {
        self.searchParams = params
        self.searchURL = SearchURLType(params)
        self.searchItems = []
    }
}

extension Search {
func getResults<each S>(_ type : repeat (each S).Type) async throws -> [(repeat (each S)?)] where repeat each S: SearchResult {
        try await withThrowingTaskGroup(of: (repeat (each S)?).self) { taskGroup in
            self.searchItems.forEach { searchitem in
                taskGroup.addTask {
                    print(searchitem.parentURL ?? "")
                        
                    let webpage = try? await searchitem.getWebpage()

                    let out = (repeat try? (each type).init(doc: webpage))
                    return out
                }
            }
            print("1")
            var searchResults : [(repeat (each S)?)] = []
            
            for try await webpage in taskGroup {
                searchResults.append(webpage)
            }
            
            return searchResults
        }
    }
private func getResultsSimple<I : SearchResult, J : SearchResult>(_ Itype : I.Type, _ Jtype : J.Type) async throws -> [(I?, J?)] {
        try await withThrowingTaskGroup(of: (I?, J?).self) { taskGroup in
            self.searchItems.forEach { searchitem in
                taskGroup.addTask {
                    print(searchitem.parentURL ?? "")
                        
                    let webpage = try? await searchitem.getWebpage()

                    let Iout = try? Itype.init(doc: webpage)
                    let Jout = try? Jtype.init(doc: webpage)
                    return (Iout, Jout)
                }
            }
            print("1")
            var searchResults : [(I?, J?)] = []
            
            for try await webpage in taskGroup {
                searchResults.append(webpage)
            }
            
            return searchResults
        }
    }   
    
public func get() async throws {
        self.searchItems = try await self.getItems()
        let test = try await self.getResults(GenericWebpage.self, GenericWebpage.self)
        //let test = try await self.getResultsSimple(GenericWebpage.self, GenericWebpage.self)
        print(test[0].0?.text ?? "nil")
    }
}
```

### Stack dump

```text
[15/15] Linking BrainLibPackageTests.xctest
Build complete! (2.04s)
Test Suite 'Selected tests' started at 2024-03-20 22:19:55.824
Test Suite 'BrainSearchTests' started at 2024-03-20 22:19:55.828
Test Case 'BrainSearchTests.testChroniclingAmerica' started at 2024-03-20 22:19:55.828
*** Signal 11: Backtracing from 0x64ea5ffa35a9...
done ***

*** Program crashed: Bad pointer dereference at 0x0000000000000000 ***

Thread 0 "BrainLibPackage":
0  0x000071684c318cce <unknown> in libc.so.6

Thread 1 crashed:
0 [system]         0x000064ea5ffa35a9 Search.getResults<A>(_:) + 553 in BrainLibPackageTests.xctest at /home/duck/AI/Brain/BrainLib/<compiler-generated>
1 [async]          0x000064ea5ffa85f0 Search.get() in BrainLibPackageTests.xctest at /home/duck/AI/Brain/BrainLib/Sources/BrainSearch/SearchGenericClass.swift:130
2 [async]          0x000064ea5ffaedd0 BrainSearchTests.testChroniclingAmerica() in BrainLibPackageTests.xctest at /home/duck/AI/Brain/BrainLib/Tests/BrainSearchTests/BrainSearchTests.swift:15
3 [async]          0x000064ea5ff57e90 implicit closure #2 in implicit closure #1 in variable initialization expression of static BrainSearchTests.__allTests__BrainSearchTests in BrainLibPackageTests.xctest at /home/duck/AI/Brain/BrainLib/.build/x86_64-unknown-linux-gnu/debug/BrainLibPackageDiscoveredTests.derived/BrainSearchTests.swift:7
4 [async] [thunk]  0x000064ea5ff586f0 partial apply for implicit closure #2 in implicit closure #1 in variable initialization expression of static BrainSearchTests.__allTests__BrainSearchTests in BrainLibPackageTests.xctest at /home/duck/AI/Brain/BrainLib/<compiler-generated>
5 [async]          0x000071684c680450 closure #1 in awaitUsingExpectation(_:) in libXCTest.so
6 [async] [thunk]  0x000071684c680cf0 partial apply for closure #1 in awaitUsingExpectation(_:) in libXCTest.so
7 [async]          0x000071684c6805e0 specialized thunk for @escaping @callee_guaranteed @Sendable @async () -> (@out A) in libXCTest.so
8 [async] [thunk]  0x000071684c681360 partial apply for specialized thunk for @escaping @callee_guaranteed @Sendable @async () -> (@out A) in libXCTest.so
9 [async] [system] 0x000071684d92a790 completeTaskWithClosure(swift::AsyncContext*, swift::SwiftError*) in libswift_Concurrency.so

Thread 2:
0  0x000071684c325e2e <unknown> in libc.so.6

Thread 3:

0  0x000071684c291117 <unknown> in libc.so.6

Thread 4:

0  0x000071684c291117 <unknown> in libc.so.6

Thread 5:

0  0x000071684c291117 <unknown> in libc.so.6

Thread 6:

0  0x000071684c291117 <unknown> in libc.so.6

Thread 7:

0  0x000071684c291117 <unknown> in libc.so.6


Registers:
rax 0x0000000000000001  1
rdx 0x0000716834003f10  00 00 00 00 00 00 00 00 b8 34 00 34 68 71 00 00  ········¸4·4hq··
rcx 0x000064ea604c9228  20 0c 00 34 68 71 00 00 70 ef 85 4d 68 71 00 00   ··4hq··pï·Mhq··
rbx 0x000064ea604c9220  02 00 00 00 00 00 00 00 20 0c 00 34 68 71 00 00  ········ ··4hq··
rsi 0x000064ea604c92a0  98 21 2c 60 ea 64 00 00 c8 91 86 4d 68 71 00 00  ·!,`êd··È··Mhq··
rdi 0x000064ea604c9218  70 91 4c 60 ea 64 00 00 02 00 00 00 00 00 00 00  p·L`êd··········
rbp 0x0000716848fbe990  f0 e9 fb 48 68 71 00 00 ad 6c 92 4d 68 71 00 00  ðéûHhq··­l·Mhq··
rsp 0x0000716848fbe8a0  08 f6 fb 48 68 71 00 00 20 0c 00 34 68 71 00 00  ·öûHhq·· ··4hq··
r8 0x0000000000000008  8
r9 0x0000000000000000  0
r10 0x0000000000000002  2
r11 0xd780d67133163a2f  15528647296850213423
r12 0x0000716848fbf608  00 00 00 00 00 00 00 00 40 8e 4c 60 ea 64 00 00  ········@·L`êd··
r13 0x0000716834000c20  70 ef 85 4d 68 71 00 00 03 00 00 00 02 00 00 00  pï·Mhq··········
r14 0x000064ea604c9200  60 91 4c 60 ea 64 00 00 f0 85 fa 5f ea 64 00 00  `·L`êd··ð·ú_êd··
r15 0x000064ea604c8e40  28 b2 94 4d 68 71 00 00 03 00 00 00 00 00 00 00  (²·Mhq··········
rip 0x000064ea5ffa35a9  4d 8b 09 4d 63 c0 4e 89 0c 07 48 83 c0 01 48 89  M··McÀN···H·À·H·
rflags 0x0000000000010206  PF

cs 0x0033  fs 0x0000  gs 0x0000
Images (48 omitted):
0x000064ea5ff1e000–0x000064ea602bdd60 <no build ID>                            BrainLibPackageTests.xctest /home/duck/AI/Brain/BrainLib/.build/x86_64-unknown-linux-gnu/debug/BrainLibPackageTests.xctest
0x000071684c200000–0x000071684c3bc341 c289da5071a3399de893d2af81d6a30c62646e1e libc.so.6                   /usr/lib/x86_64-linux-gnu/libc.so.6
0x000071684c64a000–0x000071684c6aa340 <no build ID>                            libXCTest.so                /opt/swift-5.10/usr/lib/swift/linux/libXCTest.so
0x000071684d8d0000–0x000071684d9465b0 <no build ID>                            libswift_Concurrency.so     /opt/swift-5.10/usr/lib/swift/linux/libswift_Concurrency.so

Backtrace took 0.20s
```

### Expected behavior

`

### Environment

Swift version 5.10 (swift-5.10-RELEASE)
Target: x86_64-unknown-linux-gnu

**Comment by AnthonyLatsis** [comment]
I tried completing the code to something that compiles, but could not get it to crash; there’s a fair amount of missing pieces in that reproduction. Can you provide a self-sufficient test case that compiles and crashes at run time?

**Comment by ThyOwen** [comment]
Awesome! Running this code will result in the same error.

```swift
import Foundation

public protocol SearchResult {
    init(str : String?) throws
}

public struct ResultStruct : SearchResult {
    let text : String?

    public init(str : String?) throws {
        self.text = str
    }
}

//change this class to struct and no harm done
class TestClass {
    
    var numTasks : Int

    init(numTasks : Int) {
        self.numTasks = numTasks
    }

    func getResults<each S>(_ Stype : repeat (each S).Type) async throws -> [(repeat (each S)?)] where repeat each S: SearchResult {
        try await withThrowingTaskGroup(of: (repeat (each S)?).self) { taskGroup in
            for i in 0...numTasks {
                taskGroup.addTask {

                    let out = (repeat try? (each Stype).init(str: "uhhh..."))
                    return out
                }
            }

            var searchResults : [(repeat (each S)?)] = []
                
            for try await webpage in taskGroup {
                searchResults.append(webpage)
            }
                
            return searchResults
        }
    }
}

let testClass = TestClass(numTasks : 10)



do {
    let results1 = try await testClass.getResults(ResultStruct.self)//won't crash with one. class or struct
    
    let results2 = try await testClass.getResults(ResultStruct.self, ResultStruct.self)//will crash with two or more but only with a class. not with struct
} catch {
    print("error")
}
```

**Comment by AnthonyLatsis** [comment]
> Running this code will result in the same error.

Are you sure it was the same error? Your original crash log suggests successful compilation followed by a run time crash, but this latest example [results in a compile time crash with 5.10](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1AKA7nioFSS%2BsgJ4BlRugDCqWgFcWDCQE5STgBk8BkwAOU8AI0xiEABWWNIAB1QFQnsGVw8vXySUtIEgkPCWKJj4q0wbOwEhAiZiAkzPby4/a0xbdNr6gkKwyOi4hIU6hqbs1qtR3uD%2BksH4gEorVHdiZHYOAHotgGoAFQBPRMxdw9XiXbQsXYRozFJdkl3aVCZ0XcNdzFVWRPoAHQAUg0AEE8Cxkg1dgAxVYMdBMaoMEGg1GJdwRWh4ZC7RLEVBENC0XZCTD1ZAIABKmAU7loBF2QIA7AAhVG7Tm7YKECAjS4gUkEYjBYBAgDMMMWuwICAJJgUqJZABElWCMVicbt%2Be5bLsaXSGbViLrGYKyRTqbT6YyWeywVyXphGQQfmahSKjBKYWrQY6NdjcTyCHzhbtzcLRd7pbL5QomWyOY6udYqADXapbeLldrhUnOSqlczVWDUTtKYZgKdZXh48gDAp40Rcya9YYPgxULd6ixdugBJhUfWmI2DrTGg343b80yHVyAG71XYMTz7EcAa3jgoAkoJfY7gxAVyw1wpN%2BHdruCNLp3Pk9rKmnj6fzxKc8%2BN4q7wmS2jv1R3AYXEqwIA0bUVcVnHJSlSQlbAIAAfSFY5TkFYhMBOJFdggaCEFJRYASOE5pRHQ4gJlOVUAVXYAFo4KZWJWQgdDMMZHCmBgoRFmjIFYhzEw7nQ3YWPJRlcNJc1yXWK1DVtRNv0dYVDk%2BEwmEIXYzFlfZKLMIwXwAcQJdxEggVAqEFZiMNE7DxK46MAVTG82RlDdDNWRJuRRBT712fhLjwTzdg0AEQo/M8p3kv0fOiuozzc4yAXedAXwTe0/yi6LMqdRlVizHNLNYmViEOb0bI4vDahQgjD35QUgQAJnq9wEBakLgUaxZuO8rLHXQgg1gYJ53AIGcep/UafMLUtuuTRdLiUS0wIZLcGKYkSsPYzjuMlbi%2BKZbNVt438xsdCb7z8orlKYVT1JMTAIkSJgq0C2L13ijzbwyk6FukpaCAURLEhOBEIDuh6nswLqvuiqboZ6s7kz6gaH0W61loR2HTuLItf1RegXXHZxJ32nN9kJycj1XT8Ly4DQod9VF%2B1Smd8eEtH/q4EnLpUtSCZGImRwBkC/oUCA/uNU0HMfRYdhMAQwA4RlkGIEc8M0vCBwBK5ieeHVbBnFnnTZ2SFHqrmlJ59TXX5ycAWF9nRfF4VJdTR4ndbAgpdoKgZa2MxaBJZXVY0wg8IIOWnkuFgSFOCJhqeBhaGU9XPm1wWtc7RkU71kbS2LK4kRgz7/U9EMGvq6ICWIcv6eLDhlloThYl4bwOC0UhUE4IQzAsbULg2Jl6vFHhSAITR6%2BWdcQEkYLJHFAA2eqNDn8U5/n2INHn/ROEkFvx47zheAUEANFH8fljgWAkDQSE6GichKBvxI75iUxzAIWJaYYPBgAQAgk74HQV0xBj4QAiPvCIwR6iHE4CPSBzBioAHkIjaA6GPbgvAb5sEEIgxOMC268CwHHYAzgxC0GPhg0gWAWCVnEAQqheB0KdHnLSfePwOjDU2CPYIrpG70OxBEFWxVXBYH3pGFgsD658AMMABQAA1PAmATCIOBpImQggRBiHYFIdR8glBqH3roLg%2BhKwgDfhYfQeAIjH0gMsVAiRkQUN4KgFhxARRYBsRAZY7ROgOBwgicYLR/AIj6MUUoehkipGRIEiJeRkShIGDEYxPjkTdDGG4ZoegUldGmAk%2BYSSpg9Bick3JswwmDC4N4/u2iG5Nz3vQzuHBdiqAABzzxovPSQ2oe4EFxLEAEtMaLf1/v/ZSEBcCEGeA1YeixeDoK0J1UgU96rzwBEPZei9xTinWVweeXBmTbw4LvUgrd26NKPifM%2BBCL4wEQCAXKGJLAUAgE/F%2BoRWCbFae0zp3T359IGRoIZP8/5J14JgfARB3F6H4Bo0Q4gdEwr0SodQ9CjGkBMCrRIkjakcGbic/ejTEHDUeU8KgzS2kdK6eY3pux%2BmDOGSCsZrhb70EuNMypczz6TxAFsgEWz%2BUCsFYc45pznGHysJc%2BZE9Dn1XqWc8VUrFmuNSA4SQQA). 

Reduced example:
```swift
class Class {
  func foo<each S>(_ s: repeat each S) {}
}
func test() {
  let cl = Class()
  cl.foo(true, true)
}
```
```
lib/SILGen/SILGenApply.cpp:4658: (anonymous namespace)::ClaimedParamsRef (anonymous namespace)::ParamLowering::claimParams(swift::Lowering::AbstractionPattern, ArrayRef<AnyFunctionType::Param>, const swift::ForeignInfo &): Assertion `nextSubstParamIndex == substParams.size()' failed
```

**Comment by ThyOwen** [comment]
Thank you for the response. Sorry, I wasn't clear. Yes this crash is the same as the one I experienced on my machine. Building worked. Running didn't.

`*** Program crashed: Bad pointer dereference at 0x0000000000000000 ***`

**Comment by slavapestov** [comment]
I believe this was fixed by 5f0357fdb0dfe667422ab4c5888b2ef6daba7eca which is in the release/6.0 branch but it didn't make release/5.10.
